blockchain/rolling_merkle.go

```diff
@@ -0,0 +1,149 @@
+package blockchain
+
+import (
+	"crypto/sha256"
+	"math/bits"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+)
+
+// parentHash returns the hash of the left and right hashes passed in.
+func parentHash(l, r chainhash.Hash) chainhash.Hash {
+	h := sha256.New()
+	h.Write(l[:])
+	h.Write(r[:])
+
+	buf := make([]byte, 0, 32)
+	first := h.Sum(buf)
+	h.Reset()
+	h.Write(first)
+	return *((*chainhash.Hash)(h.Sum(buf)))
+}
+
+// rollingMerkleTreeStore calculates the merkle root by only allocating O(logN)
+// memory where N is the total amount of leaves being included in the tree.
+type rollingMerkleTreeStore struct {
+	// roots are where the temporary merkle roots get stored while the
+	// merkle root is being calculated.
+	roots []chainhash.Hash
+
+	// numLeaves is the total leaves the store has processed.  numLeaves
+	// is required for the root calculation algorithm to work.
+	numLeaves uint64
+}
+
+// newRollingMerkleTreeStore returns a rollingMerkleTreeStore with the roots
+// allocated based on the passed in size.
+func newRollingMerkleTreeStore(size uint64) rollingMerkleTreeStore {
```

Comment:
oops I made a mistake and confused the default behavior of go slices with rust's vectors. It looks like this comment is entirely wrong and you should ignore it completely. On another node, I think it would be better to make the merkle store resilient to overgrowing its initial size and have the cost be performance rather than a runtime error. Thoughts?

Reply:
Actually upon further inspection it appears I was right. The `append` function in slices will do a reallocation/copy if you overrun the capacity. So the reservedSize argument is merely a performance optimization. My original comment here is correct.

Reply:
> Actually upon further inspection it appears I was right. The `append` function in slices will do a reallocation/copy if you overrun the capacity. So the reservedSize argument is merely a performance optimization. My original comment here is correct.

No the function will never allocate more than `log2(size)`, which is what the function is doing in newRollingMerkleTreeStore. You can try tweaking the benchmarking code, you'll only ever get 1 allocation per merkle root calculation.

Reply:
I'm not sure what you're saying at this point. Either it does allocate when it overruns or it doesn't. I'm asserting here that it does. It seems you agree with me. I'm not challenging your benchmarks. I'm not saying your code is slow. I'm merely documenting the behavior I am inferring from the code, which is that if you underreserve the capacity you end up using when creating a new RMTS, it will gracefully allocate more memory to accommodate that. My goal with this entire review is to make sure that the code is accessible to others who are trying to read it. This was a non-obvious property of the code that I figured I'd describe more thoroughly in the comments.

Reply:
> I'm not sure what you're saying at this point

I'm frustrated because it seems like you've not even run the code and still stating incorrect statements.

> I'm merely documenting the behavior I am inferring from the code

You've incorrectly inferred it.

> which is that if you underreserve the capacity you end up using when creating a new RMTS, it will gracefully allocate more memory to accommodate that. 

That never happens in this implementation. `newRollingMerkleTreeStore` will *always* allocate enough.

> My goal with this entire review is to make sure that the code is accessible to others who are trying to read it.

That's good! I'm stating that your change here is gonna confuse readers more because it's incorrect.

Reply:
> I'm frustrated because it seems like you've not even run the code and still stating incorrect statements.

I have run the benchmarks you cite. However, regardless of whether or not I have, the statements I am making in this thread are *independent* of the benchmark. My claims are about clarifying the role of the size reservation parameter in this private function. It so happens that you sidestep allocations by making the allocation based off of the size of the Tx vector passed into the public API. This comment, the one I am suggesting editing, is documenting the behavior of the private function, which will, in fact, reallocate if you call `.add` enough times in excess of the initial reserved size.

> That never happens in this implementation. newRollingMerkleTreeStore will always allocate enough.

Only when called by `CalcMerkleRoot`. This comment lives on the constructor/allocator of your data structure though. Even though its API is private and therefore not the concern of the public API users, anyone servicing this code may interact with the private API as they touch this code. The property I state is true of the function that the comment is attached to.

I began this investigation in response to @yyforyongyu 's question about whether or not the size can be zero. I asked myself the question "what would happen if you did call this with a size of 0?". Turns out, all the machinery will still work just fine, it'll just reallocate log(n) times. Not nearly as bad as crashing, so I actually think this is a really good design! I just was trying to specify that this function's API is capable of handling a mis-estimation of this value and the worst that will happen is an extra log(n) allocations.

> I'm stating that your change here is gonna confuse readers more because it's incorrect.

I have once again reviewed the code, because I'm new to go and it's possible I had missed something, but after a thorough investigation I once again am concluding my initial analysis to be correct. Someone else can help adjudicate this dispute if need be but in my estimation, the suggested comment for this function is accurate.

Reply:
Won't be addressing this in this PR.